### PR TITLE
ENH: manually inline PCG64DXSM code for performance.

### LINKED ISSUE #18938: PCG64DXSM performance on 64-bit Windows is poor
Surfaced on #18936

I derived the hand-"optimized" code for this case from the PCG64 code, just removing the parts that used the upper 64-bits that are missing in the cheap multiplier. I need to dive into the generated assembly to see what's going on. Issue opened for collaboration in case anyone has more assembly-adept eyeballs.

Here is the generated assembly of the 128-bit math using clang: https://godbolt.org/z/Wdj7daqaK

Here is the generated assembly from my hand-"optimized" intrinsics using MSVC: https://godbolt.org/z/dxsEWbbd8

The PCG64 hand-optimized code also manually inlined the step and output while I left those as `inline` functions here, but just to avoid making a third copy of the DXSM output function for this branch. The manual inlining for PCG64 was mostly due to the fact that it also used manual intrinsics, which shouldn't be necessary for DXSM.

### PR DESCRIPTION
**Pull request by rkern**
Closes #18938

Maybe. I don't have a 64-bit Windows test environment, but experiments in Compiler Explorer suggest this will help.

`clang` reference using `__uint128_t` arithmetic: https://godbolt.org/z/Wdj7daqaK
Compiler-inlined MSVC code: https://godbolt.org/z/dxsEWbbd8
Manually-inlined MSVC code: https://godbolt.org/z/fYWrrMfjj

@bashtage Can you check the performance on this branch?

### CODE REVIEW

File: numpy/random/src/pcg64/pcg64.h [hunk 1]
```diff
@@ -229,17 +229,6 @@ static inline void pcg_cm_step_r(pcg_state_setseq_128 *rng) {
 #endif
 }
 
-static inline uint64_t pcg_output_cm_128_64(pcg128_t state) {
```
Comment (bashtage) [inline numpy/random/src/pcg64/pcg64.h:232]: You might have seen better performance using `__forceinline` in place of `inline`.  AFAIK `__forceinline` overrides the compilers preferences, while `inline` doesn't.

**Comment by bashtage** [comment]
Looks good:

```
                      MT19937  PCG64  PCG64DXSM  Philox  SFC64
32-bit Unsigned Ints    100.0  135.7      164.3    34.5  145.2
64-bit Unsigned Ints    100.0  165.2      158.7    31.5  207.6
Uniforms                100.0  151.5      167.7    32.3  179.8
Normals                 100.0  135.6      138.7    49.1  175.5
Exponentials            100.0  155.9      158.8    34.3  173.0
Overall                 100.0  148.3      157.3    35.9  175.1
```

File: numpy/random/src/pcg64/pcg64.h (hunk 1 above)
Comment (rkern) [reply numpy/random/src/pcg64/pcg64.h:232]: We have a `#define inline __forceinline` for `_WIN32` up at the top.

**Comment by rkern** [comment]
@bashtage Thanks!

@charris This looks good to go. No behavior change, just bringing the performance back up to spec when compiled with MSVC. The organization of the C code is a bit suboptimal. Let me know if you want me to try to clean it up some.

**Comment by charris** [comment]
> The organization of the C code is a bit suboptimal.

And stylistically a bit odd. It could use a cleanup at some point, but let's put this in.

**Pull request merged**